crates/authentication/src/sys/linux/mod.rs

```diff
@@ -0,0 +1,169 @@
+            .spawn(move || {
+                let res = do_polkit_check(action_id);
+                callback(res);
+            })
+            .map_err(|_| Error::Unavailable)?;
+        Ok(())
+    }
+}
+
+static SYSTEM_CONNECTION: OnceLock<Result<Connection>> = OnceLock::new(); // Cached system bus connection
+static AUTHORITY_PROXY: OnceLock<Result<AuthorityProxyBlocking<'static>>> = OnceLock::new(); // Cached authority proxy
+
+pub(crate) fn get_system_connection() -> Result<Connection> {
+    let r: &Result<Connection> = SYSTEM_CONNECTION.get_or_init(|| {
+        Connection::system()
+            .map_err(|_| Error::Unavailable)
+    });
+
+    match r {
+        Ok(conn) => Ok(conn.clone()),
+        Err(e) => Err(e.clone()),
+    }
+}
+
+pub(crate) fn get_authority_proxy() -> Result<AuthorityProxyBlocking<'static>> {
+    let r: &Result<AuthorityProxyBlocking<'static>> = AUTHORITY_PROXY.get_or_init(|| {
+        let conn = get_system_connection()?;
+        AuthorityProxyBlocking::new(&conn).map_err(|e| {
+            let _ = e;
+            Error::Unavailable
+        })
+    });
+
+    match r {
+        Ok(p) => Ok(p.clone()),
+        Err(e) => Err(e.clone()),
+    }
+}
+
+fn do_polkit_check(action_id: &'static str) -> Result<()> {
+    // Get authority proxy, and it's cached for future use.
+    let auth = get_authority_proxy()?;
+
+    let pid = std::process::id() as u32;
+    let subject = Subject::new_for_owner(pid, None, None)
+        .map_err(|_| Error::Unavailable)?;
+
+    // If details is non-empty then the request will fail with POLKIT_ERROR_FAILED unless the process doing the check itsef is sufficiently authorized (e.g. running as uid 0).
+    let result = auth
+        .check_authorization(
+            &subject,
+            action_id,
+            &HashMap::new(), 
+            CheckAuthorizationFlags::AllowUserInteraction.into(),
+            "",
+        )
+        .map_err(|e| map_polkit_error(e.to_string()))?;
+
+    map_authorization_result(result)
+}
+
+fn map_authorization_result(result: AuthorizationResult) -> Result<()> {
+    if result.is_authorized {
+        return Ok(());
+    }
+
+    if result.details.keys().any(|k| k.contains("dismissed")) {
+        return Err(Error::UserCanceled);
+    }
+
+    if result.is_challenge {
+        // No agent available or UI interaction not possible even though we requested it.
+        return Err(Error::Unavailable);
+    }
+
+    Err(Error::Authentication)
+}
+
+fn map_polkit_error(msg: String) -> Error {
+    let m = msg.to_lowercase();
+    if m.contains("cancel") || m.contains("canceled") || m.contains("cancelled") {
+        Error::UserCanceled
+    } else if m.contains("locked") || m.contains("exhaust") || m.contains("too many") {
+        Error::Exhausted
+    } else if m.contains("unavailable") || m.contains("no agent") || m.contains("not supported") {
+        Error::Unavailable
+    } else {
+        Error::Authentication
+    }
+}
```

Comment:
surely this is not the best way of figuring out what the error is. What if the user's machine is in a different language other than english?
Typically you would use error codes or something instead of a string search.